File: VB/ZoomTrackBarControl_MiddleValue/ZoomTrackBarWithMiddleSnap.vb
Imports Microsoft.VisualBasic
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Drawing
Imports DevExpress.XtraEditors.Registrator
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraEditors.ViewInfo
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks

Namespace ZoomTrackBarControl_MiddleValue

	<UserRepositoryItem("Register")> _
	Public Class CustomRepositoryItem
		Inherits RepositoryItemZoomTrackBar
		Shared Sub New()
			Register()
		End Sub

		Public Sub New()
		End Sub

		Friend Const EditorName As String = "CustomZoomTrackBar"

		Public Shared Sub Register()
            EditorRegistrationInfo.Default.Editors.Add(New EditorClassInfo(EditorName, GetType(CustomZoomTrackBar), GetType(CustomRepositoryItem), GetType(CustomViewInfo), New ZoomTrackBarPainter(), Nothing))
		End Sub

		Public Overrides ReadOnly Property EditorTypeName() As String
			Get
				Return EditorName
			End Get
		End Property

		Private _SnapToMiddleValue As Boolean
		Public Property SnapToMiddleValue() As Boolean
			Get
				Return _SnapToMiddleValue
			End Get
			Set(ByVal value As Boolean)
				_SnapToMiddleValue = value
			End Set
		End Property

		Private _MiddleDistance As Integer
		Public Property MiddleDistance() As Integer
			Get
				If _MiddleDistance > 100 OrElse _MiddleDistance <= 0 Then
					_MiddleDistance = 10
				End If
				Return _MiddleDistance
			End Get
			Set(ByVal value As Integer)
				_MiddleDistance = value
			End Set
		End Property

		Public Overrides Sub Assign(ByVal item As RepositoryItem)
			MyBase.Assign(item)
			BeginUpdate()
			Try
				MyBase.Assign(item)
				Dim source As CustomRepositoryItem = TryCast(item, CustomRepositoryItem)
				If source Is Nothing Then
					Return
				End If
				SnapToMiddleValue = source.SnapToMiddleValue
				MiddleDistance = source.MiddleDistance
			Finally
				EndUpdate()
			End Try
		End Sub
	End Class

	<ToolboxItem(True)> _
	Public Class CustomZoomTrackBar
		Inherits ZoomTrackBarControl
		Shared Sub New()
			CustomRepositoryItem.Register()
		End Sub

		Public Sub New()
		End Sub


		Public Overrides ReadOnly Property EditorTypeName() As String
			Get
				Return CustomRepositoryItem.EditorName
			End Get
		End Property

		<DesignerSerializationVisibility(DesignerSerializationVisibility.Content)> _
		Public Shadows ReadOnly Property Properties() As CustomRepositoryItem
			Get
				Return TryCast(MyBase.Properties, CustomRepositoryItem)
			End Get
		End Property

	End Class

	Public Class CustomViewInfo
		Inherits ZoomTrackBarViewInfo
		Public Sub New(ByVal item As RepositoryItem)
			MyBase.New(item)
		End Sub

		Public Overrides Function ValueFromPoint(ByVal p As Point) As Integer
			Dim customItem As CustomRepositoryItem = CType(Me.Item, CustomRepositoryItem)
			Dim middle As Integer = If((customItem.AllowUseMiddleValue), customItem.Middle, (customItem.Minimum + customItem.Maximum) / 2)
			Dim value As Integer = MyBase.ValueFromPoint(p)
			Dim middleDistance As Integer = (customItem.Minimum + customItem.Maximum) / customItem.MiddleDistance

			If value >= middle - middleDistance AndAlso value <= middle + middleDistance Then
				Return middle
			End If
			Return value
		End Function

	End Class

End Namespace

File: CS/ZoomTrackBarControl_MiddleValue/ZoomTrackBarWithMiddleSnap.cs
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Drawing;
using DevExpress.XtraEditors.Registrator;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraEditors.ViewInfo;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ZoomTrackBarControl_MiddleValue {

    [UserRepositoryItem("Register")]
    public class CustomRepositoryItem : RepositoryItemZoomTrackBar {
        static CustomRepositoryItem() {
            Register();
        }

        public CustomRepositoryItem() { }

        internal const string EditorName = "CustomZoomTrackBar";

        public static void Register() {
            EditorRegistrationInfo.Default.Editors.Add(new EditorClassInfo(EditorName, typeof(CustomZoomTrackBar),
                typeof(CustomRepositoryItem), typeof(CustomViewInfo),
                new ZoomTrackBarPainter(), true, null, typeof(DevExpress.Accessibility.ButtonEditAccessible)));
        }

        public override string EditorTypeName {
            get { return EditorName; }
        }

        private bool _SnapToMiddleValue;
        public bool SnapToMiddleValue {
            get { return _SnapToMiddleValue; }
            set { _SnapToMiddleValue = value; }
        }

        private int _MiddleDistance;
        public int MiddleDistance {
            get {
                if (_MiddleDistance > 100 || _MiddleDistance <= 0)
                    _MiddleDistance = 10;
                return _MiddleDistance;
            }
            set {
                _MiddleDistance = value;
            }
        }

        public override void Assign(RepositoryItem item) {
            base.Assign(item);
            BeginUpdate();
            try {
                base.Assign(item);
                CustomRepositoryItem source = item as CustomRepositoryItem;
                if (source == null) return;
                SnapToMiddleValue = source.SnapToMiddleValue;
                MiddleDistance = source.MiddleDistance;
            }
            finally {
                EndUpdate();
            }
        }
    }

    [ToolboxItem(true)]
    public class CustomZoomTrackBar : ZoomTrackBarControl {
        static CustomZoomTrackBar() {
            CustomRepositoryItem.Register();
        }

        public CustomZoomTrackBar() {
        }


        public override string EditorTypeName {
            get { return CustomRepositoryItem.EditorName; }
        }

        [DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
        public new CustomRepositoryItem Properties {
            get { return base.Properties as CustomRepositoryItem; }
        }

    }

    public class CustomViewInfo : ZoomTrackBarViewInfo {
        public CustomViewInfo(RepositoryItem item)
            : base(item) {
        }

        public override int ValueFromPoint(Point p) {
            CustomRepositoryItem customItem = (CustomRepositoryItem)this.Item;
            int middle = (customItem.AllowUseMiddleValue) ? customItem.Middle : (customItem.Minimum + customItem.Maximum) / 2;
            int value = base.ValueFromPoint(p);
            int middleDistance = (customItem.Minimum + customItem.Maximum) / customItem.MiddleDistance;

            if (value >= middle - middleDistance && value <= middle + middleDistance)
                return middle;
            return value;
        }

    }

}
